File: app/src/commonMain/kotlin/ru/piterrus/aiadvent4thread/HuggingFaceClient.kt
package ru.piterrus.aiadvent4thread

import io.ktor.client.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

// Результат работы HuggingFace API
sealed class HuggingFaceResult {
    data class Success(
        val text: String, 
        val timeTaken: Long, 
        val tokensUsed: Int? = null,
        val thinkingContent: String? = null  // Для Qwen3
    ) : HuggingFaceResult()
    data class Error(val message: String) : HuggingFaceResult()
}

// Модели HuggingFace
enum class HFModel(val displayName: String, val modelPath: String, val endpoint: String) {
    STHENO(
        "L3-8B-Stheno", 
        "Sao10K/L3-8B-Stheno-v3.2",
        "https://router.huggingface.co/hf-inference/models/Sao10K/L3-8B-Stheno-v3.2"
    ),
    MINIMAX(
        "MiniMax-M2", 
        "MiniMaxAI/MiniMax-M2:novita",
        "https://router.huggingface.co/hf-inference/models/MiniMaxAI/MiniMax-M2:novita"
    ),
    QWEN2(
        "Qwen2.5-7B",
        "Qwen/Qwen2.5-7B-Instruct",
        "https://router.huggingface.co/hf-inference/models/Qwen/Qwen2.5-7B-Instruct"
    )
}

// OpenAI-совместимый формат для нового API HuggingFace
@Serializable
data class HFChatCompletionMessage(
    val role: String,
    val content: String
)

@Serializable
data class ChatCompletionRequest(
    val model: String,
    val messages: List<HFChatCompletionMessage>,
    val stream: Boolean = false,
    val max_tokens: Int? = null,
    val temperature: Double? = null,
    val top_p: Double? = null
)

@Serializable
data class ChatChoice(
    val message: HFChatCompletionMessage,
    val finish_reason: String? = null
)

@Serializable
data class ChatCompletionResponse(
    val choices: List<ChatChoice>
)

class HuggingFaceClient(private val huggingFaceToken: String) {

    private val jsonParser = Json {
        ignoreUnknownKeys = true
        isLenient = true
        prettyPrint = true
        encodeDefaults = true
    }

    private val client = HttpClient {
        install(ContentNegotiation) {
            json(jsonParser)
        }
        install(Logging) {
            logger = Logger.DEFAULT
            level = LogLevel.INFO
        }
    }

    suspend fun callStheno(prompt: String): HuggingFaceResult {
        return callModelV1(
            modelId = "Sao10K/L3-8B-Stheno-v3.2",
            modelName = "L3-8B-Stheno-v3.2",
            prompt = prompt
        )
    }
    
    suspend fun callMiniMax(prompt: String): HuggingFaceResult {
        return callModelV1(
            modelId = "MiniMaxAI/MiniMax-M2:novita",
            modelName = "MiniMax-M2",
            prompt = prompt
        )
    }
    
    suspend fun callQwen2(prompt: String, enableThinking: Boolean = true): HuggingFaceResult {
        // Формируем запрос с учетом режима thinking
        val formattedPrompt = if (enableThinking) {
            prompt
        } else {
            "$prompt /no_think"
        }
        
        val temperature = if (enableThinking) 0.6 else 0.7
        val topP = if (enableThinking) 0.95 else 0.8
        
        return callModelV1(
            modelId = "Qwen/Qwen2.5-7B-Instruct",
            modelName = "Qwen2.5-7B-Instruct",
            prompt = formattedPrompt,
            temperature = temperature,
            topP = topP,
            parseThinking = enableThinking
        )
    }
    // Новый метод для V1 API (OpenAI-совместимый)
    private suspend fun callModelV1(
        modelId: String,
        modelName: String,
        prompt: String,
        temperature: Double = 0.7,
        topP: Double = 0.9,
        parseThinking: Boolean = false
    ): HuggingFaceResult {
        return try {
            val startTime = System.currentTimeMillis()
            
            val request = ChatCompletionRequest(
                model = modelId,
                messages = listOf(
                    HFChatCompletionMessage(role = "user", content = prompt)
                ),
                stream = false,
                max_tokens = 2000,
                temperature = temperature,
                top_p = topP
            )
            
            val httpResponse: HttpResponse = client.post("https://router.huggingface.co/v1/chat/completions") {
                headers {
                    append(HttpHeaders.Authorization, "Bearer $huggingFaceToken")
                    append(HttpHeaders.ContentType, ContentType.Application.Json)
                }
                setBody(request)
            }
            
            val timeTaken = System.currentTimeMillis() - startTime
            
            when (httpResponse.status.value) {
                200 -> {
                    val responseText = httpResponse.bodyAsText()
                    println("📥 $modelName raw response: $responseText")
                    
                    try {
                        val response: ChatCompletionResponse = jsonParser.decodeFromString(responseText)
                        val fullText = response.choices.firstOrNull()?.message?.content ?: responseText
                        
                        // Парсим thinking content для Qwen, если включен режим
                        val thinkingContent: String?
                        val mainContent: String
                        
                        if (parseThinking) {
                            val thinkPattern = Regex("<think>(.*?)</think>", RegexOption.DOT_MATCHES_ALL)
                            val thinkMatch = thinkPattern.find(fullText)
                            
                            if (thinkMatch != null) {
                                thinkingContent = thinkMatch.groupValues[1].trim()
                                mainContent = fullText.replace(thinkMatch.value, "").trim()
                            } else {
                                thinkingContent = null
                                mainContent = fullText
                            }
                        } else {
                            thinkingContent = null
                            mainContent = fullText
                        }
                        
                        val tokensEstimate = mainContent.split(" ").size
                        
                        HuggingFaceResult.Success(
                            text = mainContent,
                            timeTaken = timeTaken,
                            tokensUsed = tokensEstimate,
                            thinkingContent = thinkingContent
                        )
                    } catch (e: Exception) {
                        val tokensEstimate = responseText.split(" ").size
                        HuggingFaceResult.Success(
                            text = responseText,
                            timeTaken = timeTaken,
                            tokensUsed = tokensEstimate,
                            thinkingContent = null
                        )
                    }
                }
                503 -> {
                    val errorBody = httpResponse.bodyAsText()
                    if (errorBody.contains("loading") || errorBody.contains("Loading")) {
                        HuggingFaceResult.Error("⏳ Модель $modelName загружается. Попробуйте через 20-30 секунд.")
                    } else {
                        HuggingFaceResult.Error("❌ Сервис временно недоступен (503): $errorBody")
                    }
                }
                else -> {
                    val errorBody = httpResponse.bodyAsText()
                    HuggingFaceResult.Error("❌ Ошибка ${httpResponse.status.value} для $modelName:\n$errorBody")
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            HuggingFaceResult.Error("❌ Ошибка подключения к $modelName: ${e.message}")
        }
    }
}

File: mcp-proxy/server/HuggingFaceClient.js
// =============================================================================
// HuggingFaceClient - Клиент для работы с HuggingFace API
// =============================================================================
// Использует Qwen/Qwen2.5-7B-Instruct
// =============================================================================

import dotenv from 'dotenv';
dotenv.config();

const HUGGINGFACE_API_KEY = process.env.HUGGINGFACE_API_KEY;

if (!HUGGINGFACE_API_KEY) {
    console.error('❌ Не задан HUGGINGFACE_API_KEY');
    console.error('Добавьте HUGGINGFACE_API_KEY в файл .env в папке mcp-proxy');
    process.exit(1);
}

class HuggingFaceClient {
    constructor() {
        this.modelId = 'Qwen/Qwen2.5-7B-Instruct';
        this.apiUrl = 'https://router.huggingface.co/v1/chat/completions';
    }

    /**
     * Вызвать модель L3-8B-Stheno
     * 
     * @param {Array} messages - Массив сообщений в формате {role: 'system'|'user'|'assistant', text: string}
     * @param {number} temperature - Температура генерации (по умолчанию 0.7)
     * @param {number} maxTokens - Максимальное количество токенов (по умолчанию 2000)
     * @param {Array} tools - Массив инструментов в формате MCP (опционально)
     * @returns {Promise<string>} - Ответ модели
     */
    async callModel(messages, temperature = 0.7, maxTokens = 2000, tools = null) {
        console.log('[HuggingFace] Вызов Qwen/Qwen2.5-7B-Instruct');
        console.log(`[HuggingFace] Сообщений: ${messages.length}`);
        console.log(`[HuggingFace] Temperature: ${temperature}`);
        console.log(`[HuggingFace] Tools передано: ${tools ? tools.length : 0}`);
        
        // Преобразуем формат сообщений из {role, text} в {role, content}
        const formattedMessages = messages.map(msg => ({
            role: msg.role,
            content: msg.text || msg.content
        }));
        
        const requestBody = {
            model: this.modelId,
            messages: formattedMessages,
            stream: false,
            max_tokens: maxTokens,
            temperature: temperature,
            top_p: 0.95  // Для Qwen thinking mode
        };
        
        // Добавляем tools в формате OpenAI, если они переданы
        if (tools && tools.length > 0) {
            requestBody.tools = tools.map(tool => ({
                type: 'function',
                function: {
                    name: tool.name,
                    description: tool.description,
                    parameters: tool.parameters || tool.inputSchema || {
                        type: 'object',
                        properties: {},
                        required: []
                    }
                }
            }));
            
            console.log(`[HuggingFace] ✅ Добавлено ${requestBody.tools.length} инструментов в запрос`);
            console.log(`[HuggingFace] Список: ${requestBody.tools.map(t => t.function.name).join(', ')}`);
        }

        try {
            const response = await fetch(this.apiUrl, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${HUGGINGFACE_API_KEY}`
                },
                body: JSON.stringify(requestBody)
            });

            if (!response.ok) {
                const errorText = await response.text();
                console.error('[HuggingFace] Ошибка API:', response.status, errorText);
                
                if (response.status === 503 && (errorText.includes('loading') || errorText.includes('Loading'))) {
                    throw new Error('⏳ Модель Qwen/Qwen2.5-7B-Instruct загружается. Попробуйте через 20-30 секунд.');
                }
                
                throw new Error(`HuggingFace API error: ${response.status} - ${errorText}`);
            }

            const data = await response.json();
            let responseText = data.choices[0].message.content;
            
            // Парсим и удаляем <think>...</think> блоки для Qwen thinking mode
            const thinkPattern = /<think>[\s\S]*?<\/think>/g;
            const thinkMatches = responseText.match(thinkPattern);
            if (thinkMatches) {
                console.log('[HuggingFace] 🧠 Thinking content найден, удаляем из ответа');
                responseText = responseText.replace(thinkPattern, '').trim();
            }
            
            console.log('[HuggingFace] ✅ Ответ получен');
            console.log('[HuggingFace]', responseText.substring(0, 150) + '...');
            
            return responseText;
            
        } catch (error) {
            console.error('[HuggingFace] ❌ Ошибка:', error.message);
            throw error;
        }
    }

    /**
     * Вызвать модель с системным промптом и одним пользовательским сообщением
     * (удобный метод для простых запросов)
     * 
     * @param {string} systemPrompt - Системный промпт
     * @param {string} userMessage - Сообщение пользователя
     * @param {number} temperature - Температура генерации
     * @param {number} maxTokens - Максимальное количество токенов
     * @returns {Promise<string>} - Ответ модели
     */
    async callWithPrompt(systemPrompt, userMessage, temperature = 0.7, maxTokens = 2000) {
        const messages = [
            { role: 'system', text: systemPrompt },
            { role: 'user', text: userMessage }
        ];
        
        return await this.callModel(messages, temperature, maxTokens);
    }
}

export default HuggingFaceClient;



